File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/dsl/KaleidoProtection.java
package com.tongsr.kaleido.gradle.dsl;

import javax.inject.Inject;
import org.gradle.api.Action;
import org.gradle.api.NamedDomainObjectContainer;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.SetProperty;

public abstract class KaleidoProtection {
    private final NamedDomainObjectContainer<KaleidoClassEscapeHatch> classEscapeHatches;
    private final NamedDomainObjectContainer<KaleidoResourceEscapeHatch> resourceEscapeHatches;

    @Inject
    public KaleidoProtection(ObjectFactory objects) {
        classEscapeHatches = objects.domainObjectContainer(
                KaleidoClassEscapeHatch.class,
                name -> objects.newInstance(KaleidoClassEscapeHatch.class, name));
        resourceEscapeHatches = objects.domainObjectContainer(
                KaleidoResourceEscapeHatch.class,
                name -> objects.newInstance(KaleidoResourceEscapeHatch.class, name));
    }

    public abstract SetProperty<String> getOriginalClassNames();

    public abstract SetProperty<String> getResourceNames();

    public abstract SetProperty<String> getPackagedPaths();

    public final NamedDomainObjectContainer<KaleidoClassEscapeHatch> getClassEscapeHatches() {
        return classEscapeHatches;
    }

    public final NamedDomainObjectContainer<KaleidoResourceEscapeHatch> getResourceEscapeHatches() {
        return resourceEscapeHatches;
    }

    public final void classes(
            String stableId, Action<? super KaleidoClassEscapeHatch> configure) {
        classEscapeHatches.create(stableId, configure);
    }

    public final void resources(
            String stableId, Action<? super KaleidoResourceEscapeHatch> configure) {
        resourceEscapeHatches.create(stableId, configure);
    }
}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/dsl/KaleidoProtection.kt
package com.tongsr.kaleido.gradle.dsl

import org.gradle.api.Action
import org.gradle.api.NamedDomainObjectContainer
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.SetProperty
import javax.inject.Inject

abstract class KaleidoProtection @Inject constructor(objects: ObjectFactory) {
    abstract val originalClassNames: SetProperty<String>
    abstract val resourceNames: SetProperty<String>
    abstract val packagedPaths: SetProperty<String>

    val classEscapeHatches: NamedDomainObjectContainer<KaleidoClassEscapeHatch> =
        objects.domainObjectContainer(KaleidoClassEscapeHatch::class.java) { name ->
            objects.newInstance(KaleidoClassEscapeHatch::class.java, name)
        }
    val resourceEscapeHatches: NamedDomainObjectContainer<KaleidoResourceEscapeHatch> =
        objects.domainObjectContainer(KaleidoResourceEscapeHatch::class.java) { name ->
            objects.newInstance(KaleidoResourceEscapeHatch::class.java, name)
        }

    fun classes(stableId: String, configure: Action<in KaleidoClassEscapeHatch>) {
        classEscapeHatches.create(stableId, configure)
    }

    fun resources(stableId: String, configure: Action<in KaleidoResourceEscapeHatch>) {
        resourceEscapeHatches.create(stableId, configure)
    }
}
